converter/internal/common/convert_targets.go

```diff
@@ -0,0 +1,67 @@
+package common
+
+import (
+	"github.com/grafana/agent/component/discovery"
+	"github.com/grafana/agent/pkg/river/token"
+	"github.com/grafana/agent/pkg/river/token/builder"
+)
+
+// ConvertTargets implements [builder.Tokenizer]. This allows us to set
+// component.Arguments with an implementation that can be tokenized with
+// custom behaviour for converting.
+type ConvertTargets struct {
+	Targets []discovery.Target
+}
+
+func (f ConvertTargets) RiverCapsule() {}
+func (f ConvertTargets) RiverTokenize() []builder.Token {
+	var toks []builder.Token
+	if len(f.Targets) == 0 {
+		toks = append(toks, builder.Token{Tok: token.LBRACK, Lit: ""})
+		toks = append(toks, builder.Token{Tok: token.RBRACK, Lit: ""})
+		return toks
+	}
+
+	// We are relying on each targetMap having exactly 1 target which we control
+	// from the upstream converter code. We panic below if not.
+	if len(f.Targets) > 1 {
+		toks = append(toks, builder.Token{Tok: token.LITERAL, Lit: "concat"})
+		toks = append(toks, builder.Token{Tok: token.LPAREN, Lit: ""})
+	}
+
+	for ix, targetMap := range f.Targets {
+		if len(targetMap) != 1 {
+			panic("unexpected number of targets received on a target map")
+		}
+
+		for key, target := range targetMap {
+			if key == "__address__" {
+				toks = append(toks, builder.Token{Tok: token.LBRACK, Lit: ""})
```

Comment:
I could stack all these on a single append if we prefer. More importantly, if there's a slicker way to create this whole []builder.Token let me know. I don't think it's too awful this way either though.

Reply:
Would it be easier to parse a string into tokens? The below isnt bad but just a bit verbose. Also a comment showing what the expected output would be super cool so we dont have to fully decipher the tokens meaning.

Reply:
I can add an example for sure, good idea. For the string into tokens suggestion, do we have a clear way to do that already or something new to consider?

Reply:
Agreed, I was thinking similar after matt's comment 👍

Reply:
A series of testing has been added to cover the scenarios.